pkg/controller/pkg/revision/establisher.go

```diff
@@ -0,0 +1,178 @@
+)
+
+const (
+	errAssertObj = "cannot assert object to resource.Object"
+)
+
+// An Establisher establishes control or ownership of a set of resources in the
+// API server by checking that control or ownership can be established for all
+// resources and establishing it.
+type Establisher interface {
+	Check(context.Context, []runtime.Object, resource.Object, bool) error
+	Establish(context.Context, resource.Object, bool) error
+	GetResourceRefs() []runtimev1alpha1.TypedReference
+}
+
+// APIEstablisher establishes control or ownership of resources in the API
+// server for a parent.
+type APIEstablisher struct {
+	client       client.Client
+	allObjs      []currentDesired
+	resourceRefs []runtimev1alpha1.TypedReference
+}
+
+// NewAPIEstablisher creates a new APIEstablisher.
+func NewAPIEstablisher(client client.Client) *APIEstablisher {
+	return &APIEstablisher{
+		client: client,
+	}
+}
+
+// currentDesired caches resources while checking for control or ownership so
+// that they do not have to be fetched from the API server again when control or
+// ownership is established.
+type currentDesired struct {
+	Current resource.Object
+	Desired resource.Object
+	Exists  bool
+}
+
+// GetResourceRefs returns references to resources that have had control or
+// ownership established.
+func (e *APIEstablisher) GetResourceRefs() []runtimev1alpha1.TypedReference {
+	return e.resourceRefs
+}
+
+// Check checks that control or ownership of resources can be established by
+// parent.
+func (e *APIEstablisher) Check(ctx context.Context, objs []runtime.Object, parent resource.Object, control bool) error {
+	for _, res := range objs {
+		// Assert desired object to resource.Object so that we can access its
+		// metadata.
+		d, ok := res.(resource.Object)
+		if !ok {
+			return errors.New(errAssertObj)
+		}
+
+		// Make a copy of the desired object to be populated with existing
+		// object, if it exists.
+		current := res.DeepCopyObject()
+		err := e.client.Get(ctx, types.NamespacedName{Name: d.GetName(), Namespace: d.GetNamespace()}, current)
+		if resource.IgnoreNotFound(err) != nil {
+			return err
+		}
+
+		// If resource does not already exist, we must attempt to dry run create
+		// it.
+		if kerrors.IsNotFound(err) {
+			// Add to objects as not existing.
+			e.allObjs = append(e.allObjs, currentDesired{
+				Desired: d,
+				Current: nil,
+				Exists:  false,
+			})
+			if err := e.create(ctx, d, parent, control, client.DryRunAll); err != nil {
+				return err
+			}
+			continue
+		}
+
+		c, ok := current.(resource.Object)
+		if !ok {
+			return errors.New(errAssertObj)
+		}
+
+		// Add to objects as existing.
+		e.allObjs = append(e.allObjs, currentDesired{
+			Desired: d,
+			Current: c,
+			Exists:  true,
+		})
+
+		if err := e.update(ctx, c, d, parent, control, client.DryRunAll); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Establish establishes control or ownership of resources in the package.
```

Comment:
Could be worth clarifying that the caller must call Check before calling Establish.